extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,654 @@
+    devicePlayer.setAttribute('height', device.height);
+    const storyA = devicePlayer.querySelector('a');
+    storyA.textContent = 'Story 1';
+    storyA.href = this.storyUrl_;
+    setStyles(devicePlayer, {
+      width: device.width + 'px',
+      height: device.height + 'px',
+    });
+    const playerImpl = new AmpStoryPlayer(this.win, devicePlayer);
+    playerImpl.load();
+    setStyles(
+      deviceLayout.querySelector('.i-amphtml-story-dev-tools-device-screen'),
+      {
+        height: device.deviceHeight
+          ? device.deviceHeight + 'px'
+          : 'fit-content',
+      }
+    );
+    device.player = devicePlayer;
+    return deviceLayout;
+  }
+
+  /**
+   * @private
+   * @param {DeviceInfo} deviceSpecs
+   */
+  addDevice_(deviceSpecs) {
+    const deviceLayout = this.buildDeviceLayout_(deviceSpecs, this.storyUrl_);
+    deviceSpecs.element = deviceLayout;
+    this.element.appendChild(deviceLayout);
+    deviceSpecs.chip = this.addDeviceChip_(deviceSpecs);
+    this.devices_.push(deviceSpecs);
+    this.onLayoutMeasure();
+    setStyles(deviceLayout, {'opacity': '1'});
+    setTimeout(() => {
+      setStyles(deviceLayout, {
+        'transition': 'transform ease-out 0.15s',
+      });
+    }, 150);
+  }
+
+  /**
+   * @private
+   * @param {DeviceInfo} device
+   */
+  removeDevice_(device) {
+    device.chip.setAttribute('inactive', '');
+    setTimeout(() => {
+      device.chip.remove();
+    }, 200);
+    this.devices_ = this.devices_.filter((d) => d != device);
+    device.element.remove();
+    this.onLayoutMeasure();
+  }
+
+  /**
+   * @param {DeviceInfo} device
+   * @return {!Element} the device chip
+   */
+  addDeviceChip_(device) {
+    const deviceChip = buildDeviceChipTemplate(this.element);
+    deviceChip.querySelector('span').textContent = device.name;
+    deviceChip.device = device;
+    this.element
+      .querySelector('.i-amphtml-story-dev-tools-device-chips')
+      .appendChild(deviceChip);
+    return deviceChip;
+  }
+
+  /**
+   * @private
+   */
+  updateDevicesInHash_() {
+    const devicesStringRepresentation = this.devices_
+      .map((device) => {
+        if (device.custom) {
+          return (
+            `${device.width}:${device.height}` +
+            (device.name ? ':' + device.name : '')
+          );
+        }
+        return device.name.toLowerCase().replace(/[^a-z0-9]/gi, '');
+      })
+      .join(';');
+    this.element.setAttribute('devices', devicesStringRepresentation);
+    updateHash({'devices': devicesStringRepresentation}, this.win);
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    const layoutBox = this.getLayoutBox();
+    let sumDeviceWidths = 0;
+    let maxDeviceHeights = 0;
+    this.devices_.forEach((deviceSpecs) => {
+      sumDeviceWidths += deviceSpecs.width;
+      maxDeviceHeights = Math.max(
+        maxDeviceHeights,
+        deviceSpecs.deviceHeight || deviceSpecs.height + 100
+      );
+    });
```

Comment:
You could remove a lot of this math by putting the devices in a flex container.
That flex container could then be scaled instead of each individual element.
CSS could then also handle positioning instead of transforming each element.

Sketch PR:
https://github.com/mszylkowski/amphtml/compare/mszylkowski:ampstory_devtools_tabs_preview...processprocess:flex-container#

Reply:
I think it might not work with the device transitions properly, and we'd need `width: fit contents` to make the container as wide as needed (otherwise it works). I've also seen that the players (and the stories) look blurred if they are not positioned in absolute units and scaled using transformZ, does that happen to your branch?

If we want to keep the device animations when you add or remove the devices, I'm not sure that we could use the flexbox model to make it work.

Reply:
Yes we would remove device transitions with this approach. The transitions are currently janky so we need to identify ways to show the user that a new device has appeared but make it as smooth as possible. 

We might want to remove any decorative animations so we can focus on honing transitions that educate the user on actions. 

The rotation/opacity fade on the + to - might be a candidate for removal.
![Nov-25-2020 14-23-10](https://user-images.githubusercontent.com/3860311/100272913-eb227280-2f29-11eb-821a-6d6ea42402a1.gif)

Reply:
Transforming one wrapper element instead of each player element could help this. 
[codepen sketch](https://codepen.io/philipbell/pen/jOMEePQ)

Reply:
Thought of that, but using that approach would make it impossible animate the removal of devices properly if the device removed is not the last one (which could happen without the dialog, when the devices are 100% visible). I don't think it would be a good idea to get rid of the animations in that case because it's hard to track the devices when they jump around. On that sense I'd say that it's even better to have 2-3 frames of a laggy animation than no transitions, but:

I tried to optimize the animations (using only transform for GPU acceleration) on tuesday/wednesday and it improved, and tested on the Mac and personal PC laptop and both are smooth. From what I'm seeing, it does work smooth when adding and removing devices even with 8-9 devices, but we'll limit the number to 4. Updated the demo link ([here](https://stampinteractivitybackend.web.app/examples/amp-story/interactives.html#development=1))

Reply:
This optimization work is great but there's noticeable frame clipping when adding and removing devices.

![Dec-02-2020 10-21-30](https://user-images.githubusercontent.com/3860311/100892670-91b9c680-3488-11eb-9683-1b5786e724a2.gif)

<img width="575" alt="Screen Shot 2020-12-02 at 10 22 34 AM" src="https://user-images.githubusercontent.com/3860311/100892665-8ff00300-3488-11eb-81b6-4a5e1cbff65c.png">

Reply:
Yeah I think this is as performant as it will get, because the player loading will jank the UI no matter what... It's loading a full story with potentially thousands of nodes in the iframe, so makes sense.

I still feel like this result is better than no animation at all, as it's fast enough on good systems to show smoothly, and when it stutters it will still hint the user what device got added or removed; so functionally it works as intended. 

Also users won't be adding or removing devices very often, as the default devices will represent most ratios. Being in the background, I don't think the UX will suffer much from the animation not being 60fps, specifically if users don't run the animations nonstop like we do to analyze them.